packages/typespec-test/test/contentsafety_modular/generated/typespec-ts/src/api/pagingHelpers.ts

```diff
@@ -0,0 +1,183 @@
+        nextPageLink: nextLink,
+      };
+    },
+    byPage: (settings?: TPageSettings) => {
+      const { continuationToken } = settings ?? {};
+      return getPageAsyncIterator(pagedResult, {
+        pageLink: continuationToken,
+      });
+    },
+  };
+  return getPagedAsyncIterator(pagedResult);
+}
+
+/**
+ * returns an async iterator that iterates over results. It also has a `byPage`
+ * method that returns pages of items at once.
+ *
+ * @param pagedResult - an object that specifies how to get pages.
+ * @returns a paged async iterator that iterates over results.
+ */
+
+function getPagedAsyncIterator<
+  TElement,
+  TPage = TElement[],
+  TPageSettings extends PageSettings = PageSettings
+>(
+  pagedResult: PagedResult<TElement, TPage, TPageSettings>
+): PagedAsyncIterableIterator<TElement, TPage, TPageSettings> {
+  const iter = getItemAsyncIterator<TElement, TPage, TPageSettings>(
+    pagedResult
+  );
+  return {
+    next() {
+      return iter.next();
+    },
+    [Symbol.asyncIterator]() {
+      return this;
+    },
+    byPage:
+      pagedResult?.byPage ??
+      ((settings?: TPageSettings) => {
+        const { continuationToken } = settings ?? {};
+        return getPageAsyncIterator(pagedResult, {
+          pageLink: continuationToken,
+        });
+      }),
+  };
+}
+
+async function* getItemAsyncIterator<
+  TElement,
+  TPage,
+  TPageSettings extends PageSettings
+>(
+  pagedResult: PagedResult<TElement, TPage, TPageSettings>
+): AsyncIterableIterator<TElement> {
+  const pages = getPageAsyncIterator(pagedResult);
+  for await (const page of pages) {
+    yield* page as unknown as TElement[];
+  }
+}
+
+async function* getPageAsyncIterator<
+  TElement,
+  TPage,
+  TPageSettings extends PageSettings
+>(
+  pagedResult: PagedResult<TElement, TPage, TPageSettings>,
+  options: {
+    pageLink?: string;
+  } = {}
+): AsyncIterableIterator<ContinuablePage<TElement, TPage>> {
+  const { pageLink } = options;
+  let response = await pagedResult.getPage(
+    pageLink ?? pagedResult.firstPageLink
+  );
+  if (!response) {
+    return;
+  }
+  let result = response.page as ContinuablePage<TElement, TPage>;
+  result.continuationToken = response.nextPageLink;
+  yield result;
+  while (response.nextPageLink) {
+    response = await pagedResult.getPage(response.nextPageLink);
+    if (!response) {
+      return;
+    }
+    result = response.page as ContinuablePage<TElement, TPage>;
+    result.continuationToken = response.nextPageLink;
+    yield result;
+  }
+}
+
+/**
+ * Gets for the value of nextLink in the body
+ */
+function getNextLink(body: unknown, nextLinkName?: string): string | undefined {
+  if (!nextLinkName) {
+    return undefined;
+  }
+
+  const nextLink = (body as Record<string, unknown>)[nextLinkName];
+
+  if (typeof nextLink !== "string" && typeof nextLink !== "undefined") {
+    throw new RestError(
+      `Body Property ${nextLinkName} should be a string or undefined`
+    );
+  }
```

Comment:
What if it is set to null? I was just burnt by this 😅

Reply:
Goot catch! I would add the possible value also `null`.